fireplace-swing/src/main/java/io/github/bric3/fireplace/flamegraph/FlameGraphPainter.java

```diff
@@ -647,37 +609,30 @@ public void stopHover() {
         hoveredFrame = null;
     }
 
-    private void paintFrameText(T node, Graphics2D g2, Font font, double targetWidth, Consumer<String> textConsumer) {
+    private String calculateFrameText(Graphics2D g2, Font font, double targetWidth, FrameBox<T> frame) {
         var metrics = g2.getFontMetrics(font);
 
-        // don't use stream to avoid allocations during painting
-        for (Function<T, String> nodeToTextCandidate : nodeToTextCandidates) {
-            var textCandidate = nodeToTextCandidate.apply(node);
-            var textBounds = metrics.getStringBounds(textCandidate, g2);
-            if (textBounds.getWidth() > targetWidth) {
-                continue;
-            }
-            textConsumer.accept(textCandidate);
-            return;
-        }
-        var textBounds = metrics.getStringBounds(LONG_TEXT_PLACEHOLDER, g2);
-        if (!(textBounds.getWidth() > targetWidth)) {
-            textConsumer.accept(LONG_TEXT_PLACEHOLDER);
-        }
-    }
-
-    private static void paintRootFrameText(String text, Graphics2D g2, Font font, double targetWidth, Consumer<String> textConsumer) {
-        var metrics = g2.getFontMetrics(font);
+        String text = nodeToTextProvider.frameString(frame.actualNode, frame.stackDepth == 0);
 
         var textBounds = metrics.getStringBounds(text, g2);
-        if (!(textBounds.getWidth() > targetWidth)) {
-            textConsumer.accept(text);
+        if (textBounds.getWidth() <= targetWidth) {
+            return text;
         }
-        textBounds = metrics.getStringBounds(LONG_TEXT_PLACEHOLDER, g2);
-        if (!(textBounds.getWidth() > targetWidth)) {
-            textConsumer.accept(LONG_TEXT_PLACEHOLDER);
+
+        var avgPerChar = metrics.charWidth('A');
+        var numOfChars = Math.floor(targetWidth / avgPerChar) - 2;
+        if (numOfChars < 0) {
+            return null;
+        }
+
+        var shortedString = text.substring(0, (int) numOfChars) + LONG_TEXT_PLACEHOLDER;
+        textBounds = metrics.getStringBounds(shortedString, g2);
+        if (textBounds.getWidth() <= targetWidth) {
+            return shortedString;
         }
+
```

Comment:
I'm not really keen on this change yet, I find this quite useful to be able to add a shorter text representation if available.

I do like the idea of truncating text, however. It feels like at least part of this behavior should be configurable, but as you and I discussed we don't want to expose `g2`, (font metrics are ok, but it requires `g2`).